### PR TITLE
Add PhysicalMemoryLimit capabilities for /proc/meminfo OS's like Linux.

### LINKED ISSUE #1426: MemoryCache needs a way to handle physical memory pressure on Linux
We need to design a way to allow MemoryCache to handle physical memory pressure. This could be something similar to what we do on Windows (just with a different lookup table to translate the memory % usage to the actual pressure - for this we will likely need dotnet/corefx#25617 to be implemented first). Or this could be some additional extensibility point to allow user code to communicate memory pressure back to cache.

### PR DESCRIPTION
**Pull request by StephenMolloy**
Enable MemCache PhysicalMemoryLimit on Linux.

This fix uses procfs instead of [System.GC.GetGCMemoryInfo](https://docs.microsoft.com/en-us/dotnet/api/system.gc.getgcmemoryinfo?view=netcore-3.0)
as was suggested in #24299. The choice to use /proc/meminfo is because
that method should work on pretty much any major Linux distro and it
gives us "accurate" information at the time of query. Using
GetGCMemoryInfo is using potentially stale information, though it might
be worth considering it as a last-ditch fallback for other Unix variants if
procfs isn't available?

Fix #1426

### CODE REVIEW

**Comment by jkotas** [comment]
`/proc/meminfo` won't give you the right limit for containers.

File: src/libraries/System.Runtime.Caching/src/System/Runtime/Caching/PhysicalMemoryMonitor.Unix.cs [hunk 1]
```diff
@@ -2,14 +2,81 @@
...
+        // says dwMemoryLoad represents "approximate percentage of physical memory that is in use".
+        // Exact precision or perscribing to a particular definition of "physical memory that is in use"
+        // is not of utmost importance here. Having roughly analagous behavior on Windows and Unix is the goal.
+        private static bool ReadMemInfo(out ulong total, out ulong available)
+        {
+            bool totalValid = false;
+            bool availableValid = false;
+            total = available = 0;
+
+            if (File.Exists("/proc/meminfo"))
```
Comment (mconnew) [inline src/libraries/System.Runtime.Caching/src/System/Runtime/Caching/PhysicalMemoryMonitor.Unix.cs:60]: It's possible the file could exist but the current process doesn't have read permission. You should wrap the file read in a try/catch and make it behave the same way as if the file didn't exist.

File: src/libraries/System.Runtime.Caching/tests/System.Runtime.Caching/MemoryCacheTest.cs [hunk 2]
```diff
@@ -39,11 +39,36 @@
...
+        {
+            get
+            {
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                {
+                    return true;
+                }
+
+                // Other OS's support if they have /proc/meminfo with "MemAvailable"
+                if (File.Exists("/proc/meminfo"))
```
Comment (mconnew) [inline src/libraries/System.Runtime.Caching/tests/System.Runtime.Caching/MemoryCacheTest.cs:60]: Same comment as earlier one, consider file existing but not being readable

**Comment by mconnew** [comment]
Regarding the comment by @jkotas, when in a container environment there are 2 scenarios. First, no memory quota on the container. In this case you probably still care about the host value here. Second, when you do have a memory quota on the container then you probably don't need to use this feature. This feature is to dynamically respond to other processes on the machine which are consuming memory. In the container model you only have the one process so there's no "other" to consider. In this case you can likely use a fixed size cache.

**Comment by StephenMolloy** [comment]
> `/proc/meminfo` won't give you the right limit for containers.

@jkotas , how would you suggest handling the container case then? I've been digging through the sources behind GC.GetGCMemoryInfo() and I'm not quite sure the numbers there are consistent for a container scenario either. In [GCToOSInterface::GetMemoryStatus](https://github.com/dotnet/runtime/blob/master/src/coreclr/gc/unix/gcenv.unix.cpp#L1309), the oddly named `restricted_limit` parameter is only non-zero when there is not restricted container limit. In which case, lines 1317-1326 calculate straightforward numbers as expected.

But if there is a restricetd limit, it looks like the call to [GetAvailablePhysicalMemory](https://github.com/dotnet/runtime/blob/master/src/coreclr/gc/unix/gcenv.unix.cpp#L1190) on line 1329 will check /proc/meminfo@MemAvailable like we do here, or fall back to sysconf. Either way, this is not a container-aware stat. The call to [GetPhysicalMemoryLimit](https://github.com/dotnet/runtime/blob/master/src/coreclr/gc/unix/gcenv.unix.cpp#L1138) on line 1334 does appear to get a total memory size that _is_ container aware. So it's possible to get an 'available' value that is larger than 'total' and not even get a 'memory_load' calculation as a result.

Later, [when gathering these stats into a struct to return to managed land](https://github.com/dotnet/runtime/blob/master/src/coreclr/gc/gc.cpp#L42126), the memory load bytes field is calculated by multiplying the percentage load by total_physical_memory... which _might_ be container aware? I can't tell for sure. But if it was container aware, then there wouldn't be a need to choose between it and 'heap_hard_limit' for totalAvailableMemoryBytes then.

To summarize, it doesn't look like GC.GetGCMemoryInfo is consistently container aware either. I'm not sure how I would use it in this case. Maybe I'm missing something? Or perhaps you have another solution in mind?

**Comment by StephenMolloy** [comment]
Then again, as @mconnew mentioned - the original idea behind this monitor was to be a good citizen among other processes on the machine. In the container scenario, a cache of any significant size is likely to be part of the only citizen that matters within that container, so using a container aware monitor to be a good citizen within the container doesn't seem necessary.

On the other hand, using a host-aware monitor might be useful to help the container be a good citizen among other containers/processes on the host - especially if the cache-containing-container was not created with a quota.

**Comment by jkotas** [comment]
> the oddly named restricted_limit parameter is only non-zero when there is not restricted container limit.

It is the other way: restricted_limit is true when the container has a limit.

I do not have a strong opinion on this. If you believe that it does not make for this feature to work in containers, feel free to ignore my comment. It would be nice to document it somewhere.

**Comment by davidfowl** [comment]
Seems like this should be using GC APIs to get memory information. The array pool does a similar thing and responds to memory pressure. Why not do a similar thing here

**Comment by StephenMolloy** [comment]
Ok, I'm looking at [GCHeap::Initialize](https://github.com/dotnet/runtime/blob/master/src/coreclr/gc/gc.cpp#L39749) again, and I see where `total_physical_mem` is being set near the beginning to either a .Net configured limit, or a os/container-configured limit, or the actual total amount of RAM installed. I'm still having difficulty seeing how I get a percentage load back from GCMemoryInfo though.

Internally the GC is tracking load as a percentage, but before exposing it, it multiplies that by `total_physical_mem`. That's fine. But how can I know what that value is to undo that calculation? It looks like `total_physical_mem` was exposed by GCMemoryInfo.TotalAvailableMemoryBytes previously... but now that field will either expose the `total_physical_mem` or it will expose 75% of that value in a restricted memory scenario. Or thats what it looks like to me in lines 39825-39843. `heap_hard_limit` will be set if it is configured as a percentage through .Net config, or it will be set to 75% of total memory in a "restricted" scenario. The end result is that I can't reliably use GCMemoryInfo.TotalAvailableMemoryBytes to convert back to an accurate memory load percentage.

A [comment ](https://github.com/dotnet/coreclr/pull/25437#discussion_r299810957) by @jkotas from the PR that introduced the hard limit to GCMemoryInfo suggests that in container scenarios, caches should care about this GC hard limit and not the actual memory load relative to the container memory limit. Perhaps that is true when managing array pools, and maybe that is a relevant consideration when the cache tuning being discussed is only concerned about it's own heap usage bumping into limits. But this cache monitor does not serve this purpose. It would rather know the status of memory use in the entire system/container so it can get out of other peoples way - not so it can get out of it's own way. And that's the way this cache has been doing it on Windows for years going back to .Net 4.X. Is there another way to get `total_physical_mem` from GC that I'm missing?

**Comment by StephenMolloy** [comment]
Updated PR to use GCMemoryInfo on non-Windows. It's not perfect, but the scenarios in which GCMemoryInfo do not match what we do on windows are restricted-memory scenarios... in which case there is not much need for this "good citizen in an environment shared with other concerns" memory monitor. This leaves the complexity of working with /proc/meminfo or sysctl or CLR memory limits to the GC system which has been doing this for years.

File: src/libraries/System.Runtime.Caching/src/System/Runtime/Caching/PhysicalMemoryMonitor.Unix.cs [hunk 3]
```diff
@@ -2,14 +2,53 @@
...
+        private long lastGCIndex;
+
         protected override int GetCurrentPressure()
         {
-            return 0;
+            // Get stats from GC. Try to get them to refresh if they haven't been updated since our last check.
+            GCMemoryInfo memInfo = GC.GetGCMemoryInfo();
+            if (memInfo.Index <= lastGCIndex)
+            {
+                GC.Collect(0); // A quick, ephemeral Gen 0 collection
```
Comment (mconnew) [inline src/libraries/System.Runtime.Caching/src/System/Runtime/Caching/PhysicalMemoryMonitor.Unix.cs:35]: Consider using the overload which takes a GCCollectionMode and pass GCCollectionMode.Optimized. This allows the GC to determine when only a trivial amount of allocation has happened since the last GC and to skip it. In other words, if the memory situation hasn't have changed an appreciable amount, it skips the GC.

**Review by mconnew — APPROVED** [review]
I'm a little concerned about forcing GC every time, so I suggested the more polite GC collect api. Not requiring that change, I'll leave it to you to decide.

**Review by HongGit — APPROVED** [review]
